File: sdks/php/tests/Node/NodeRegistryTest.php
<?php

declare(strict_types=1);

namespace Blok\Nanoservice\Tests\Node;

use Blok\Nanoservice\Node\NodeHandler;
use Blok\Nanoservice\Node\NodeRegistry;
use Blok\Nanoservice\Types\Context;
use Blok\Nanoservice\Types\ExecutionRequest;
use Blok\Nanoservice\Types\NodeConfig;
use Blok\Nanoservice\Testing\MockContext;
use PHPUnit\Framework\TestCase;

final class NodeRegistryTest extends TestCase
{
    private function createTestNode(mixed $returnValue): NodeHandler
    {
        return new class($returnValue) implements NodeHandler {
            public function __construct(private readonly mixed $data) {}

            public function execute(Context $ctx, array $config): mixed
            {
                return $this->data;
            }
        };
    }

    public function testRegisterAndGet(): void
    {
        $registry = new NodeRegistry('1.0.0');
        $node = $this->createTestNode(['msg' => 'hello']);

        $registry->register('test', $node);

        $this->assertNotNull($registry->get('test'));
        $this->assertNull($registry->get('missing'));
    }

    public function testNodeNames(): void
    {
        $registry = new NodeRegistry('1.0.0');
        $registry->register('alpha', $this->createTestNode(null));
        $registry->register('beta', $this->createTestNode(null));

        $names = $registry->nodeNames();
        $this->assertCount(2, $names);
        $this->assertContains('alpha', $names);
        $this->assertContains('beta', $names);
    }

    public function testCountAndIsEmpty(): void
    {
        $registry = new NodeRegistry('1.0.0');
        $this->assertTrue($registry->isEmpty());
        $this->assertSame(0, $registry->count());

        $registry->register('test', $this->createTestNode(null));
        $this->assertFalse($registry->isEmpty());
        $this->assertSame(1, $registry->count());
    }

    public function testExecuteSuccess(): void
    {
        $registry = new NodeRegistry('1.0.0');
        $registry->register('test', $this->createTestNode(['msg' => 'hello']));

        $req = new ExecutionRequest(
            node: new NodeConfig(name: 'test'),
            context: MockContext::create()->build(),
        );

        $result = $registry->execute($req);

        $this->assertTrue($result->success);
        $this->assertSame(['msg' => 'hello'], $result->data);
        $this->assertNotNull($result->metrics);
        $this->assertNotNull($result->metrics->durationMs);
        $this->assertGreaterThanOrEqual(0, $result->metrics->durationMs);
    }

    public function testExecuteNotFound(): void
    {
        $registry = new NodeRegistry('1.0.0');

        $req = new ExecutionRequest(
            node: new NodeConfig(name: 'missing'),
            context: MockContext::create()->build(),
        );

        $result = $registry->execute($req);

        $this->assertFalse($result->success);
        $this->assertNotNull($result->errors);
        $this->assertStringContainsString('not found', $result->errors['message']);
    }

    public function testExecuteHandlesException(): void
    {
        $failingNode = new class implements NodeHandler {
            public function execute(Context $ctx, array $config): mixed
            {
                throw new \RuntimeException('boom');
            }
        };

        $registry = new NodeRegistry('1.0.0');
        $registry->register('fail', $failingNode);

        $req = new ExecutionRequest(
            node: new NodeConfig(name: 'fail'),
            context: MockContext::create()->build(),
        );

        $result = $registry->execute($req);

        $this->assertFalse($result->success);
        $this->assertStringContainsString('boom', $result->errors['message']);
    }

    public function testHealth(): void
    {
        $registry = new NodeRegistry('2.0.0');
        $registry->register('a', $this->createTestNode(null));
        $registry->register('b', $this->createTestNode(null));

        $health = $registry->health();

        $this->assertSame('healthy', $health->status);
        $this->assertSame('2.0.0', $health->version);
        $this->assertCount(2, $health->nodesLoaded);
    }
}

File: sdks/ruby/test/node/node_registry_test.rb
# frozen_string_literal: true

require_relative "../test_helper"

# Simple test node that returns configured data.
class TestNode < Nanoservice::Node::NodeHandler
  def initialize(data = { "ok" => true })
    super()
    @data = data
  end

  def execute(_ctx, _config)
    @data
  end
end

# Node that raises an error.
class ErrorNode < Nanoservice::Node::NodeHandler
  def execute(_ctx, _config)
    raise Nanoservice::Errors::NodeError.validation("bad input")
  end
end

class NodeRegistryTest < Minitest::Test
  def setup
    @registry = Nanoservice::Node::NodeRegistry.new("1.0.0")
  end

  def test_register_and_get
    @registry.register("test", TestNode.new)

    assert_instance_of TestNode, @registry.get("test")
    assert_nil @registry.get("missing")
  end

  def test_node_names
    @registry.register("beta", TestNode.new)
    @registry.register("alpha", TestNode.new)

    names = @registry.node_names
    assert_equal %w[alpha beta], names
  end

  def test_execute_success
    @registry.register("test", TestNode.new({ "msg" => "hello" }))

    ctx     = Nanoservice::Testing::MockContext.new.build
    request = Nanoservice::Types::ExecutionRequest.new(
      node:    Nanoservice::Types::NodeConfig.new(name: "test"),
      context: ctx
    )
    result = @registry.execute(request)

    assert result.success
    assert_equal "hello", result.data["msg"]
    assert result.metrics
    assert result.metrics.duration_ms > 0
  end

  def test_execute_not_found
    ctx     = Nanoservice::Testing::MockContext.new.build
    request = Nanoservice::Types::ExecutionRequest.new(
      node:    Nanoservice::Types::NodeConfig.new(name: "missing"),
      context: ctx
    )
    result = @registry.execute(request)

    refute result.success
    assert_match(/not found/, result.errors["message"])
  end

  def test_execute_error_handling
    @registry.register("error", ErrorNode.new)

    ctx     = Nanoservice::Testing::MockContext.new.build
    request = Nanoservice::Types::ExecutionRequest.new(
      node:    Nanoservice::Types::NodeConfig.new(name: "error"),
      context: ctx
    )
    result = @registry.execute(request)

    refute result.success
    assert result.metrics
    assert result.metrics.duration_ms >= 0
  end

  def test_health
    @registry.register("a", TestNode.new)
    @registry.register("b", TestNode.new)

    health = @registry.health

    assert_equal "healthy", health.status
    assert_equal "1.0.0", health.version
    assert_equal 2, health.nodes_loaded.length
    assert_includes health.nodes_loaded, "a"
    assert_includes health.nodes_loaded, "b"
  end

  def test_middleware_applied
    log_entries = []
    middleware = Class.new(Nanoservice::Middleware::Middleware) do
      define_method(:initialize) do |entries|
        super()
        @entries = entries
      end

      define_method(:wrap) do |handler|
        entries = @entries
        ->(ctx, config) {
          entries << "before"
          result = handler.call(ctx, config)
          entries << "after"
          result
        }
      end
    end

    @registry.use(middleware.new(log_entries))
    @registry.register("test", TestNode.new)

    ctx     = Nanoservice::Testing::MockContext.new.build
    request = Nanoservice::Types::ExecutionRequest.new(
      node:    Nanoservice::Types::NodeConfig.new(name: "test"),
      context: ctx
    )
    result = @registry.execute(request)

    assert result.success
    assert_equal %w[before after], log_entries
  end
end
